File: mini-agent-harness/typescript/agent/extensionRegistry.ts
import type { CapabilityDefinition } from '../capabilityProjection.ts'

export type ExtensionComponentKind = 'skill' | 'command' | 'tool' | 'agent'
export type ExtensionTrust = 'local-trusted' | 'signed' | 'untrusted'

export type ExtensionSourceIdentity = Readonly<{
  marketplace: string
  locator: string
  plugin: string
  version: string
}>

export type ExtensionComponent = Readonly<{
  kind: ExtensionComponentKind
  name: string
  description: string
  priority?: number
  deferred?: boolean
}>

export type ExtensionBundle = Readonly<{
  namespace: string
  source: ExtensionSourceIdentity
  trust: ExtensionTrust
  signature?: string
  components: readonly ExtensionComponent[]
}>

export type RegisteredExtension = Readonly<{
  qualifiedName: string
  bundleKey: string
  namespace: string
  source: ExtensionSourceIdentity
  component: ExtensionComponent
}>

export type ExtensionSnapshot = Readonly<{
  revision: number
  entries: readonly RegisteredExtension[]
}>

export type ExtensionConflict = Readonly<{
  qualifiedName: string
  bundleKeys: readonly string[]
}>

export type ExtensionPublication =
  | Readonly<{ ok: true; snapshot: ExtensionSnapshot }>
  | Readonly<{ ok: false; revision: number; conflicts: readonly ExtensionConflict[] }>

export type ExtensionTrace = Readonly<{
  action: 'publish' | 'reject-conflict' | 'acquire' | 'release'
  revision: number
  bundleCount: number
  conflictCount: number
}>

export interface ExtensionTrustPolicy {
  assertAllowed(bundle: ExtensionBundle): void
}

export class SignatureTrustPolicy implements ExtensionTrustPolicy {
  assertAllowed(bundle: ExtensionBundle): void {
    if (bundle.trust === 'untrusted') throw new Error('untrusted extension bundle')
    if (bundle.trust === 'signed' && !bundle.signature?.trim()) {
      throw new Error('signed extension bundle requires signature evidence')
    }
  }
}

export type ExtensionLease = Readonly<{
  bundleKey: string
  qualifiedName: string
  acquiredRevision: number
  isValid(): boolean
  release(): void
}>

export class StaleExtensionRevisionError extends Error {}

export class ExtensionRegistry {
  #revision = 0
  #entries: readonly RegisteredExtension[] = Object.freeze([])
  #activeBundleKeys = new Set<string>()
  #leaseSequence = 0
  readonly #leases = new Map<number, { released: boolean }>()
  readonly #trustPolicy: ExtensionTrustPolicy
  readonly #trace: ExtensionTrace[] = []

  constructor(trustPolicy: ExtensionTrustPolicy = new SignatureTrustPolicy()) {
    this.#trustPolicy = trustPolicy
  }

  snapshot(): ExtensionSnapshot {
    return deepFreeze({ revision: this.#revision, entries: this.#entries })
  }

  publish(expectedRevision: number, bundles: readonly ExtensionBundle[]): ExtensionPublication {
    if (expectedRevision !== this.#revision) {
      throw new StaleExtensionRevisionError(
        `stale extension revision ${expectedRevision}; current=${this.#revision}`,
      )
    }
    const frozenBundles = bundles.map(bundle => validateAndFreezeBundle(bundle, this.#trustPolicy))
    const candidates = frozenBundles.flatMap(bundle => bundle.components.map(component =>
      deepFreeze({
        qualifiedName: `${bundle.namespace}:${component.name}`,
        bundleKey: sourceKey(bundle.source),
        namespace: bundle.namespace,
        source: bundle.source,
        component,
      }),
    ))
    const conflicts = findConflicts(candidates)
    if (conflicts.length > 0) {
      this.#record('reject-conflict', frozenBundles.length, conflicts.length)
      return deepFreeze({ ok: false, revision: this.#revision, conflicts })
    }

    this.#entries = Object.freeze([...candidates].sort((left, right) =>
      compareText(left.qualifiedName, right.qualifiedName)
      || compareText(left.bundleKey, right.bundleKey),
    ))
    this.#activeBundleKeys = new Set(frozenBundles.map(bundle => sourceKey(bundle.source)))
    this.#revision += 1
    this.#record('publish', frozenBundles.length, 0)
    return deepFreeze({ ok: true, snapshot: this.snapshot() })
  }

  acquire(snapshot: ExtensionSnapshot, qualifiedName: string): ExtensionLease {
    const entry = snapshot.entries.find(candidate => candidate.qualifiedName === qualifiedName)
    if (!entry) throw new Error(`extension is not visible in snapshot: ${qualifiedName}`)
    if (!this.#activeBundleKeys.has(entry.bundleKey)) {
      throw new Error(`extension version is no longer active: ${qualifiedName}`)
    }
    const leaseId = ++this.#leaseSequence
    const state = { released: false }
    this.#leases.set(leaseId, state)
    this.#record('acquire', this.#activeBundleKeys.size, 0)
    return Object.freeze({
      bundleKey: entry.bundleKey,
      qualifiedName,
      acquiredRevision: snapshot.revision,
      isValid: () => !state.released,
      release: () => {
        if (state.released) return
        state.released = true
        this.#leases.delete(leaseId)
        this.#record('release', this.#activeBundleKeys.size, 0)
      },
    })
  }

  traces(): readonly ExtensionTrace[] {
    return deepFreeze(this.#trace.map(item => ({ ...item })))
  }

  toCapabilityDefinitions(snapshot: ExtensionSnapshot): readonly CapabilityDefinition[] {
    return deepFreeze(snapshot.entries.map(entry => ({
      name: entry.qualifiedName,
      description: entry.component.description,
      source: 'plugin' as const,
      priority: entry.component.priority ?? 50,
      ...(entry.component.deferred ? { deferred: true } : {}),
    })))
  }

  #record(action: ExtensionTrace['action'], bundleCount: number, conflictCount: number): void {
    this.#trace.push(Object.freeze({ action, revision: this.#revision, bundleCount, conflictCount }))
  }
}

function validateAndFreezeBundle(
  bundle: ExtensionBundle,
  trustPolicy: ExtensionTrustPolicy,
): ExtensionBundle {
  requireIdentifier(bundle.namespace, 'extension namespace')
  for (const [label, value] of Object.entries(bundle.source)) requireText(value, `source ${label}`)
  if (bundle.components.length === 0) throw new Error('extension bundle must contain components')
  trustPolicy.assertAllowed(bundle)
  const names = new Set<string>()
  for (const component of bundle.components) {
    requireIdentifier(component.name, 'component name')
    requireText(component.description, 'component description')
    if (names.has(component.name)) throw new Error(`duplicate component in bundle: ${component.name}`)
    names.add(component.name)
  }
  return deepFreeze(structuredClone(bundle))
}

function findConflicts(entries: readonly RegisteredExtension[]): readonly ExtensionConflict[] {
  const owners = new Map<string, Set<string>>()
  for (const entry of entries) {
    const keys = owners.get(entry.qualifiedName) ?? new Set<string>()
    keys.add(entry.bundleKey)
    owners.set(entry.qualifiedName, keys)
  }
  return deepFreeze([...owners.entries()]
    .filter(([, keys]) => keys.size > 1)
    .map(([qualifiedName, keys]) => ({ qualifiedName, bundleKeys: [...keys].sort(compareText) }))
    .sort((left, right) => compareText(left.qualifiedName, right.qualifiedName)))
}

function sourceKey(source: ExtensionSourceIdentity): string {
  return `${source.marketplace}::${source.locator}::${source.plugin}::${source.version}`
}

function requireIdentifier(value: string, label: string): void {
  requireText(value, label)
  if (!/^[A-Za-z][A-Za-z0-9_.-]*$/.test(value)) {
    throw new Error(`${label} must use the portable ASCII identifier form`)
  }
}

function requireText(value: string, label: string): void {
  if (!value.trim()) throw new Error(`${label} must not be empty`)
}

function compareText(left: string, right: string): number {
  return left < right ? -1 : left > right ? 1 : 0
}

function deepFreeze<T>(value: T): T {
  if (value !== null && typeof value === 'object' && !Object.isFrozen(value)) {
    for (const nested of Object.values(value as Record<string, unknown>)) deepFreeze(nested)
    Object.freeze(value)
  }
  return value
}

File: mini-agent-harness/python/extension_registry.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Callable, Literal, Protocol
import re

ExtensionComponentKind = Literal["skill", "command", "tool", "agent"]
ExtensionTrust = Literal["local-trusted", "signed", "untrusted"]
TraceAction = Literal["publish", "reject-conflict", "acquire", "release"]


@dataclass(frozen=True)
class ExtensionSourceIdentity:
    marketplace: str
    locator: str
    plugin: str
    version: str


@dataclass(frozen=True)
class ExtensionComponent:
    kind: ExtensionComponentKind
    name: str
    description: str
    priority: float = 50
    deferred: bool = False


@dataclass(frozen=True)
class ExtensionBundle:
    namespace: str
    source: ExtensionSourceIdentity
    trust: ExtensionTrust
    components: tuple[ExtensionComponent, ...]
    signature: str | None = None


@dataclass(frozen=True)
class RegisteredExtension:
    qualified_name: str
    bundle_key: str
    namespace: str
    source: ExtensionSourceIdentity
    component: ExtensionComponent


@dataclass(frozen=True)
class ExtensionSnapshot:
    revision: int
    entries: tuple[RegisteredExtension, ...]


@dataclass(frozen=True)
class ExtensionConflict:
    qualified_name: str
    bundle_keys: tuple[str, ...]


@dataclass(frozen=True)
class ExtensionPublication:
    ok: bool
    revision: int
    snapshot: ExtensionSnapshot | None = None
    conflicts: tuple[ExtensionConflict, ...] = ()


@dataclass(frozen=True)
class ExtensionTrace:
    action: TraceAction
    revision: int
    bundle_count: int
    conflict_count: int


class ExtensionTrustPolicy(Protocol):
    def assert_allowed(self, bundle: ExtensionBundle) -> None: ...


class SignatureTrustPolicy:
    def assert_allowed(self, bundle: ExtensionBundle) -> None:
        if bundle.trust == "untrusted":
            raise ValueError("untrusted extension bundle")
        if bundle.trust == "signed" and not (bundle.signature or "").strip():
            raise ValueError("signed extension bundle requires signature evidence")


class StaleExtensionRevisionError(RuntimeError):
    pass


class ExtensionLease:
    def __init__(
        self,
        bundle_key: str,
        qualified_name: str,
        acquired_revision: int,
        release_callback: Callable[[], None],
    ) -> None:
        self.bundle_key = bundle_key
        self.qualified_name = qualified_name
        self.acquired_revision = acquired_revision
        self._release_callback = release_callback
        self._released = False

    @property
    def valid(self) -> bool:
        return not self._released

    def release(self) -> None:
        if self._released:
            return
        self._released = True
        self._release_callback()


class ExtensionRegistry:
    def __init__(self, trust_policy: ExtensionTrustPolicy | None = None) -> None:
        self._revision = 0
        self._entries: tuple[RegisteredExtension, ...] = ()
        self._active_bundle_keys: frozenset[str] = frozenset()
        self._trust_policy = trust_policy or SignatureTrustPolicy()
        self._traces: list[ExtensionTrace] = []
        self._lease_sequence = 0
        self._leases: set[int] = set()

    def snapshot(self) -> ExtensionSnapshot:
        return ExtensionSnapshot(self._revision, self._entries)

    def publish(
        self, expected_revision: int, bundles: tuple[ExtensionBundle, ...]
    ) -> ExtensionPublication:
        if expected_revision != self._revision:
            raise StaleExtensionRevisionError(
                f"stale extension revision {expected_revision}; current={self._revision}"
            )
        for bundle in bundles:
            _validate_bundle(bundle, self._trust_policy)
        candidates = tuple(
            RegisteredExtension(
                f"{bundle.namespace}:{component.name}",
                _source_key(bundle.source),
                bundle.namespace,
                bundle.source,
                component,
            )
            for bundle in bundles
            for component in bundle.components
        )
        conflicts = _find_conflicts(candidates)
        if conflicts:
            self._record("reject-conflict", len(bundles), len(conflicts))
            return ExtensionPublication(False, self._revision, conflicts=conflicts)

        self._entries = tuple(
            sorted(candidates, key=lambda item: (item.qualified_name, item.bundle_key))
        )
        self._active_bundle_keys = frozenset(
            _source_key(bundle.source) for bundle in bundles
        )
        self._revision += 1
        self._record("publish", len(bundles), 0)
        snapshot = self.snapshot()
        return ExtensionPublication(True, self._revision, snapshot=snapshot)

    def acquire(
        self, snapshot: ExtensionSnapshot, qualified_name: str
    ) -> ExtensionLease:
        entry = next(
            (
                candidate
                for candidate in snapshot.entries
                if candidate.qualified_name == qualified_name
            ),
            None,
        )
        if entry is None:
            raise ValueError(f"extension is not visible in snapshot: {qualified_name}")
        if entry.bundle_key not in self._active_bundle_keys:
            raise ValueError(f"extension version is no longer active: {qualified_name}")
        self._lease_sequence += 1
        lease_id = self._lease_sequence
        self._leases.add(lease_id)
        self._record("acquire", len(self._active_bundle_keys), 0)

        def release() -> None:
            self._leases.discard(lease_id)
            self._record("release", len(self._active_bundle_keys), 0)

        return ExtensionLease(
            entry.bundle_key, qualified_name, snapshot.revision, release
        )

    def traces(self) -> tuple[ExtensionTrace, ...]:
        return tuple(self._traces)

    def capability_definitions(
        self, snapshot: ExtensionSnapshot
    ) -> tuple[tuple[str, str, float, bool], ...]:
        return tuple(
            (
                entry.qualified_name,
                entry.component.description,
                entry.component.priority,
                entry.component.deferred,
            )
            for entry in snapshot.entries
        )

    def _record(
        self, action: TraceAction, bundle_count: int, conflict_count: int
    ) -> None:
        self._traces.append(
            ExtensionTrace(
                action, self._revision, bundle_count, conflict_count
            )
        )


def _validate_bundle(
    bundle: ExtensionBundle, trust_policy: ExtensionTrustPolicy
) -> None:
    _require_identifier(bundle.namespace, "extension namespace")
    for label, value in (
        ("marketplace", bundle.source.marketplace),
        ("locator", bundle.source.locator),
        ("plugin", bundle.source.plugin),
        ("version", bundle.source.version),
    ):
        _require_text(value, f"source {label}")
    if not bundle.components:
        raise ValueError("extension bundle must contain components")
    trust_policy.assert_allowed(bundle)
    names: set[str] = set()
    for component in bundle.components:
        _require_identifier(component.name, "component name")
        _require_text(component.description, "component description")
        if component.name in names:
            raise ValueError(f"duplicate component in bundle: {component.name}")
        names.add(component.name)


def _find_conflicts(
    entries: tuple[RegisteredExtension, ...]
) -> tuple[ExtensionConflict, ...]:
    owners: dict[str, set[str]] = {}
    for entry in entries:
        owners.setdefault(entry.qualified_name, set()).add(entry.bundle_key)
    return tuple(
        ExtensionConflict(name, tuple(sorted(keys)))
        for name, keys in sorted(owners.items())
        if len(keys) > 1
    )


def _source_key(source: ExtensionSourceIdentity) -> str:
    return (
        f"{source.marketplace}::{source.locator}::"
        f"{source.plugin}::{source.version}"
    )


def _require_identifier(value: str, label: str) -> None:
    _require_text(value, label)
    if re.fullmatch(r"[A-Za-z][A-Za-z0-9_.-]*", value) is None:
        raise ValueError(f"{label} must use the portable ASCII identifier form")


def _require_text(value: str, label: str) -> None:
    if not value.strip():
        raise ValueError(f"{label} must not be empty")
